### PR TITLE
DAOS-4449 control: Set up fault domain for server

### PR DESCRIPTION
**Pull request by kjacque**
- Use the fault path from the server config file, or use the host
  name as a single-level fault domain.
- Send the control server's fault domain as part of the Join request
  to the MS. Nothing is done with that yet.
- Add some useful functions for dealing with fault domains.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

File: src/control/system/faultdomain_test.go [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+			common.CmpErr(t, tc.expErr, err)
+
+			if diff := cmp.Diff(result, tc.expResult, cmp.AllowUnexported(FaultDomain{})); diff != "" {
+				t.Fatalf("(-want, +got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestSystem_FaultDomain_String(t *testing.T) {
```
Comment (mjmac) [inline src/control/system/faultdomain_test.go:165]: Not necessarily anything wrong with doing it this way, but you could combine the previous suite with this one and just compare the output of .String() to the input string to make sure that they match. Just a bit less maintenance overhead.

**Review by tanabarr — DISMISSED** [review]

**Review by mjmac — COMMENTED** [review]
Looks good overall, just a few comments. In particular, would prefer not to expand the ext stuff.

File: src/control/system/faultdomain_test.go (hunk 1 above)
Comment (kjacque) [reply src/control/system/faultdomain_test.go:165]: Hm, but in some cases we normalize the input string. I guess we could add an expStr to the test case struct in the previous test. I think keeping them separate may be a bit easier to understand, though.

File: src/control/server/instance.go [hunk 2]
```diff
@@ -161,12 +169,10 @@ func (srv *IOServerInstance) removeSocket() error {
...
 }
 
-// setRank determines the instance rank and sends a SetRank dRPC request
-// to the IOServer.
-func (srv *IOServerInstance) setRank(ctx context.Context, ready *srvpb.NotifyReadyReq) error {
+func (srv *IOServerInstance) determineRank(ctx context.Context, ready *srvpb.NotifyReadyReq) (system.Rank, error) {
 	superblock := srv.getSuperblock()
 	if superblock == nil {
-		return errors.New("nil superblock in setRank()")
+		return system.NilRank, errors.New("nil superblock in setRank()")
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/server/instance.go:175]: Good catch. Fixed.

File: src/control/server/external.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ type External interface {
 	getAbsInstallPath(string) (string, error)
 	checkSudo() (bool, string)
 	getHistory() []string
+	getHostname() (string, error)
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/server/external.go:37]: Ah, got it. I have an idea of how to rework the tests.

File: src/control/system/faultdomain_test.go (hunk 1 above)
Comment (mjmac) [reply src/control/system/faultdomain_test.go:165]: Yeah that's fine. The test is already written and I wouldn't expect these tests to need a lot of maintenance.

**Review by tanabarr — APPROVED** [review]

**Review by mjmac — APPROVED** [review]

**Pull request merged**